indexer/services/roundtable/src/tasks/aggregate-trading-rewards.ts

```diff
@@ -1,63 +1,230 @@
 }
 
-async function getTradingRewardDataToProcessInterval(): Promise<Interval> {
-  const latestBlock: BlockFromDatabase = await BlockTable.getLatest();
+export class AggregateTradingReward {
+  period: TradingRewardAggregationPeriod;
 
-  // TODO(IND-499): Setup AggregateTradingRewardsProcessedCache for start time and add end time
-  return {
-    start: DateTime.fromISO(latestBlock.time),
-    end: DateTime.fromISO(latestBlock.time),
-  };
-}
+  constructor(period: TradingRewardAggregationPeriod) {
+    this.period = period;
+  }
 
-async function getTradingRewardDataToProcess(
-  _interval: Interval,
-): Promise<TradingRewardFromDatabase[]> {
-  // TODO: Implement
-  return Promise.resolve([]);
-}
+  async runTask(): Promise<void> {
+    const interval: Interval | undefined = await this.getTradingRewardDataToProcessInterval();
+    if (interval === undefined) {
+      logger.info({
+        at: 'aggregate-trading-rewards#runTask',
+        message: 'No interval to aggregate trading rewards',
+      });
+      return;
+    }
+    logger.info({
+      at: 'aggregate-trading-rewards#runTask',
+      message: 'Generated interval to aggregate trading rewards',
+      start: interval.start.toISO(),
+      end: interval.end.toISO(),
+    });
 
-function sortTradingRewardData(
-  _tradingRewardData: TradingRewardFromDatabase[],
-): SortedTradingRewardData {
-  // TODO: Implement
-  return {};
-}
+    const tradingRewardData:
+    TradingRewardFromDatabase[] = await this.getTradingRewardDataToProcess(interval);
+    const sortedTradingRewardData: SortedTradingRewardData = this.sortTradingRewardData(
+      tradingRewardData,
+    );
+    await this.updateTradingRewardsAggregation(sortedTradingRewardData);
+    await AggregateTradingRewardsProcessedCache.setProcessedTime(
+      this.period,
+      interval.end.toISO(),
+      redisClient,
+    );
+  }
+
+  /**
+   * Gets the interval of time to aggregate trading rewards for.
+   * If there are no blocks in the database, then do not process any data.
+   * If There is no processedTime in the cache, then delete the latest month of data,
+   * and reprocess that data or start from block 1.
+   * If the processedTime is not null, and blocks exist in the database, then process up to the
+   * next hour of data.
+   */
+  async getTradingRewardDataToProcessInterval(): Promise<Interval | undefined> {
+    let latestBlock: BlockFromDatabase;
+    try {
+      latestBlock = await BlockTable.getLatest();
+    } catch (e) {
+      logger.info({
+        at: 'aggregate-trading-rewards#getTradingRewardDataToProcessInterval',
+        message: 'Unable to aggregate trading rewards because there are no blocks in the database.',
+      });
+      return;
+    }
+
+    const processedTime:
+    IsoString | null = await AggregateTradingRewardsProcessedCache.getProcessedTime(
+      this.period,
+      redisClient,
+    );
+    if (processedTime === null) {
+      await this.deleteIncompleteAggregatedTradingReward();
```

Comment:
It's a little confusing that this `getTradingRewardDataToProcessInterval` function has the side effect of deleting trading reward aggregation. Up to you whether you want to refactor this part out.

Reply:
moved it out, good call out